chore: add upgrade guide for v.1.5

marcelstanley opened the pull request:
Relates to #432. Requirement for v1.5 

----
Evidence that the steps were executed on devnet and that inputs/outputs from v1.4.0 match the ones in v1.5.0 after the upgrade is complete:
- Commands executed: [console_log.txt](https://github.com/user-attachments/files/16117169/console_log.txt)
- anvil/node logs: [node_upgrade_log.txt](https://github.com/user-attachments/files/16117170/node_upgrade_log.txt)

marcelstanley committed:
chore: add upgrade guide for v.1.5

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+
+## How to upgrade
+
+Application owners may decide to redeploy their applications, upgrading their instances of the Cartesi Rollups Node to release `v.1.5.0`.
```
```suggestion
Application owners may decide to redeploy their applications, upgrading their instances of the Cartesi Rollups Node to release `v1.5.0`.
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+## Steps to replace an Application's _History_
+
+> [!CAUTION]
+> Instances of the [History](https://github.com/cartesi/rollups-contracts/blob/v1.4.0/onchain/rollups/contracts/history/History.sol) contract from [rollups-contracts v.1.4.0](https://github.com/cartesi/rollups-contracts/releases/tag/v1.4.0) may be used simultaneously by several applications through their associated _Authority_ instance.
```
```suggestion
> Instances of the [History](https://github.com/cartesi/rollups-contracts/blob/v1.4.0/onchain/rollups/contracts/history/History.sol) contract from [rollups-contracts v1.4.0](https://github.com/cartesi/rollups-contracts/releases/tag/v1.4.0) may be used simultaneously by several applications through their associated _Authority_ instance.
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+First, the address of the new _History_ instance should be calculated.
+
+> [!NOTE]
+> It's recommended to use the deterministic functions available in the Rollups Contracts.
```
```suggestion
> It's recommended to use the deterministic deployment functions available in the Rollups Contracts.
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+> [!IMPORTANT]
+> The commands below assumes that the following environment variables have been previous defined:
+>
+> - `CONTRACT_OWNER_ADDRESS`: address of the wallet used by the application owner
```
This environment variable is not necessary.

marcelstanley replied:
Removed at 1fbb3de

marcelstanley replied:
All the other suggestions were accepted as well.

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+> - `HISTORY_FACTORY_ADDRESS`: address of a valid _History Factory_ instance
+> - `AUTHORITY_ADDRESS`: address of the _Authority_ instance being used by the application
+
+In order to calculate the new address, the owner may call function [`calculateHistoryAddress(address, bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L35-L38).
```
It doesn't matter who calls this view function.

```suggestion
In order to calculate the address of a new history contract, one can call the view function [`calculateHistoryAddress(address,bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L35-L38).
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+    --trace --verbose \
+    $HISTORY_FACTORY_ADDRESS \
+    "calculateHistoryAddress(address,bytes32)(address)" \
+    $CONTRACT_OWNER_ADDRESS \
```
You want the owner of the History contract to be the Authority contract, making it the only authorized caller of the `submitClaim` function.

```suggestion
    $AUTHORITY_ADDRESS \
```

guidanoli commented on docs/upgrade_to_1_5.md:
````diff
@@ -0,0 +1,119 @@
+    --rpc-url "$RPC_URL"
+```
+
+Assuming the command execution is successfull, set the resulting address as a new environment variable `NEW_HISTORY_ADDRESS`, to be used later on.
````
```suggestion
Assuming the command execution is successfull, it should print out the address of the to-be-created History contract. You may store this address in the environment variable `NEW_HISTORY_ADDRESS`, to be used later on.
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+    --mnemonic "$MNEMONIC" \
+    $HISTORY_FACTORY_ADDRESS \
+    "newHistory(address,bytes32)(History)" \
+    $CONTRACT_OWNER_ADDRESS \
```
```suggestion
    $AUTHORITY_ADDRESS \
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+
+Assuming the command execution is successfull, set the resulting address as a new environment variable `NEW_HISTORY_ADDRESS`, to be used later on.
+
+After that, the new instance of _History_ may be created by calling function [`newHistory(address, bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L24-L27) as follows:
```
```suggestion
After that, the new instance of _History_ may be created by calling function [`newHistory(address,bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L24-L27) as follows:
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+> [!IMPORTANT]
+> The command below assumes that the environment variables set in the previous step are available, as well as  `NEW_HISTORY_ADDRESS`, which must contain the address of the new _History_.
+
+To replace the _History_ used by the _Authority_, call [`setHistory(_history)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/consensus/authority/Authority.sol#L65) as follows:
```
```suggestion
To replace the _History_ used by the _Authority_, call [`setHistory(address)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/consensus/authority/Authority.sol#L65) as follows:
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,119 @@
+    "$AUTHORITY_ADDRESS" \
+    "setHistory(address)" \
+    "$NEW_HISTORY_ADDRESS" \
+    "$SALT" \
```
The function doesn't expect a salt as parameter.

```suggestion
```

guidanoli reviewed (commented):
It might be worth noting that the new History will contain no claims, and that they will need to be submitted by the Authority again, one by one.

Might be interesting also to run these `cast` commands locally, on a local fork or local devnet, and show an example of the expected output, so that users can know if they're following the migration guide correctly.

Also, would be interesting to show how these read and write functions can be called on Etherscan. Might be a more friendly UI for less seasoned web3 developers. We have contracts verified for the main networks, and one can easily call read-only functions (such as `calculateHistoryAddress`, and, after connecting their browser wallets, submit state-changing txs (such as `newHistory` and `setHistory`). In this case, screenshots might be useful.

marcelstanley commented:
> Might be interesting also to run these `cast` commands locally, on a local fork or local devnet, and show an example of the expected output, so that users can know if they're following the migration guide correctly.

Sure, that was the plan!

Here's evidence that the steps were executed on devnet and that inputs/outputs from v1.4.0 match the ones in v1.5.0 after the upgrade is complete:
- Commands executed: [console_log.txt](https://github.com/user-attachments/files/16117169/console_log.txt)
- anvil/node logs: [node_upgrade_log.txt](https://github.com/user-attachments/files/16117170/node_upgrade_log.txt)

marcelstanley commented:
> It might be worth noting that the new History will contain no claims, and that they will need to be submitted by the Authority again, one by one.

Added [note at a2dab18](https://github.com/cartesi/rollups-node/blob/a2dab18c8199912156c6c4a89da5b76f1ee46088/docs/upgrade_to_1_5.md?plain=1#L45-L47)

marcelstanley commented:
> It might be worth noting that the new History will contain no claims, and that they will need to be submitted by the Authority again, one by one.
> 


Added [note at a2dab18](https://github.com/cartesi/rollups-node/blob/a2dab18c8199912156c6c4a89da5b76f1ee46088/docs/upgrade_to_1_5.md?plain=1#L45-L47)

> Might be interesting also to run these `cast` commands locally, on a local fork or local devnet, and show an example of the expected output, so that users can know if they're following the migration guide correctly.
> 

Sure, that was always the plan! I've just added that to the PR description above.

> Also, would be interesting to show how these read and write functions can be called on Etherscan. Might be a more friendly UI for less seasoned web3 developers. We have contracts verified for the main networks, and one can easily call read-only functions (such as `calculateHistoryAddress`, and, after connecting their browser wallets, submit state-changing txs (such as `newHistory` and `setHistory`). In this case, screenshots might be useful.
>

That would be a nice addition. Perhaps we should move this whole guide to the wiki, then. 🤔 

Thanks a lot for your help, @guidanoli!

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,124 @@
+> A redeployment will create a new instance of the application from scratch.
+> All previous inputs, outputs and claims will remain associated to the previous application address, funds included.
+
+Alternatively, if the application owner wants to keep the application inputs, they may choose to replace the _History_ being used by the application's _Authority_ instead.
```
```suggestion
Alternatively, the _Authority_ owner may choose to replace the _History_ being used by the application's _Authority_ instead. This path allows inputs, outputs and locked funds to remain unchanged.
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,124 @@
+
+> [!NOTE]
+> A new _History_ will contain no claims.
+> Once the Rollups Node is restarted with a new _History_, all previous claims will be submitted again, one-by-one, by the configured _Authority_, incurring in extra cost for the application owner.
```
It's healthy to exercise a separation between the application owner (who can change the application's consensus contract) and _Authority_ owner (who can submit claims and change the current _History_).

```suggestion
> Once the Rollups Node is restarted with a new _History_, all previous claims will be submitted again, one-by-one, by the configured _Authority_, incurring in extra cost for the _Authority_ owner.
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,124 @@
+> [!IMPORTANT]
+> All commands below assume that the following environment variables have been previous defined:
+>
+> - `SALT`: salt to be used by the deterministic deployment functions
```
Some users might not know what a salt is. Let's be more explicit here for those users.


```suggestion
> - `SALT`: a random 32-byte value to be used by the deterministic deployment functions
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,124 @@
+>
+> - `SALT`: salt to be used by the deterministic deployment functions
+> - `RPC_URL`: the RPC endpoint to be used
+> - `MNEMONIC`: mnemonic phrase for the application owner's wallet (other wallet options may be used)
```
You need the mnemonic of the _Authority_ owner that to call the `setHistory` function on the _Authority_ contract. For deploying the new _History_ contract, however, any account can be used. So, might be good to rename this as `AUTHORITY_OWNER_MNEMONIC`, and have `MNEMONIC` reference any account with enough funds.

```suggestion
> - `MNEMONIC`: mnemonic phrase for the _Authority_ owner's wallet (other wallet options may be used)
```

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,124 @@
+> - `SALT`: salt to be used by the deterministic deployment functions
+> - `RPC_URL`: the RPC endpoint to be used
+> - `MNEMONIC`: mnemonic phrase for the application owner's wallet (other wallet options may be used)
+> - `HISTORY_FACTORY_ADDRESS`: address of a valid _History Factory_ instance
```
Would be nice to state here that for all supported networks (Ethereum, Optimism, Arbitrum, Base + respective Sepolia-based tesnets), the `HistoryFactory` is deployed at `0x1f158b5320BBf677FdA89F9a438df99BbE560A26`.

marcelstanley replied:
Done at 92b8b68f1862e. Please close this conversation of it's OK

marcelstanley commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,124 @@
+Application owners may decide to redeploy their applications, upgrading their instances of the Cartesi Rollups Node to release `v1.5.0`.
+This is the simplest way to perform the upgrade.
+
+In order to do so, just update the application configuration considering [what has changed](#what-has-changed) and upgrade the version of the Cartesi Rollups Node as usual.
```
Done at 92b8b68f1862e. Please close this conversation of it's OK

guidanoli commented on docs/upgrade_to_1_5.md:
```diff
@@ -0,0 +1,130 @@
+
+In order to do so, just update the application configuration considering [what has changed](#what-has-changed) and upgrade the version of the Cartesi Rollups Node as usual.
+
+For more details about hot to deploy an application, please refer to the [deployment process](https://docs.cartesi.io/cartesi-rollups/1.3/deployment/introduction/#deployment-process) options available in the Cartesi documentation.
```
```suggestion
For more details about how to deploy an application, please refer to the [deployment process](https://docs.cartesi.io/cartesi-rollups/1.3/deployment/introduction/#deployment-process) options available in the Cartesi documentation.
```

guidanoli reviewed (approved):
LGTM
